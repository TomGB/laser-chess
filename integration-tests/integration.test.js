const mockUserInput = jest.fn();
const mockDrawPieces = jest.fn();

jest.mock('../get-user-input', () => mockUserInput);
jest.mock('../draw-pieces', () => mockDrawPieces);

const gameLoop = require('../game-loop');
const setupGame = require('../setup-game');

const mockMove = move => mockUserInput.mockImplementationOnce(() => move);

test('getUserInput is called', () => {
    [
        { start: { x: 4 , y: 0 }, move: { x: 4, y: 1 } },
        { start: { x: 3 , y: 2 }, move: { x: 3, y: 3 } },
        { start: { x: 0 , y: 0 }, rotate: 'l' },
    ].forEach(mockMove);

    const game = setupGame();
    expect(game).toMatchSnapshot('Initial game');
    const outcome = gameLoop(game);

    expect(outcome).toEqual('Finished! Game won by white')

    mockDrawPieces.mock.calls.forEach(([ call ]) =>
        expect(call).toMatchSnapshot('Board state each turn')
    );
    mockUserInput.mock.calls.forEach(([ call ]) => 
        expect(call).toMatchSnapshot('Each turn')
    );
});