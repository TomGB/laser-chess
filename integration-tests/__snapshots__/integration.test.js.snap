// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserInput is called: Board state each turn 1`] = `
Array [
  Array [
    Object {
      "colour": "red",
      "rotation": "R",
      "rotation2": undefined,
      "type": "laser",
      "x": 0,
      "y": 0,
    },
    null,
    null,
    null,
    null,
    Object {
      "colour": "red",
      "rotation": undefined,
      "rotation2": undefined,
      "type": "king",
      "x": 5,
      "y": 0,
    },
    Object {
      "colour": "red",
      "rotation": "D",
      "rotation2": undefined,
      "type": "block",
      "x": 6,
      "y": 0,
    },
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": undefined,
      "type": "corner",
      "x": 7,
      "y": 0,
    },
    null,
    null,
  ],
  Array [
    null,
    null,
    Object {
      "colour": "red",
      "rotation": "DL",
      "rotation2": undefined,
      "type": "corner",
      "x": 2,
      "y": 1,
    },
    null,
    Object {
      "colour": "red",
      "rotation": "D",
      "rotation2": undefined,
      "type": "block",
      "x": 4,
      "y": 1,
    },
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    Object {
      "colour": "red",
      "rotation": "UR",
      "rotation2": undefined,
      "type": "corner",
      "x": 0,
      "y": 3,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "DL",
      "rotation2": undefined,
      "type": "corner",
      "x": 2,
      "y": 3,
    },
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": undefined,
      "type": "corner",
      "x": 3,
      "y": 3,
    },
    Object {
      "colour": "red",
      "rotation": "UR",
      "rotation2": "DL",
      "type": "mirror",
      "x": 4,
      "y": 3,
    },
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": "LU",
      "type": "mirror",
      "x": 5,
      "y": 3,
    },
    null,
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": undefined,
      "type": "corner",
      "x": 7,
      "y": 3,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": undefined,
      "type": "corner",
      "x": 9,
      "y": 3,
    },
  ],
  Array [
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": undefined,
      "type": "corner",
      "x": 0,
      "y": 4,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": undefined,
      "type": "corner",
      "x": 2,
      "y": 4,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": "RD",
      "type": "mirror",
      "x": 4,
      "y": 4,
    },
    Object {
      "colour": "white",
      "rotation": "DL",
      "rotation2": "UR",
      "type": "mirror",
      "x": 5,
      "y": 4,
    },
    null,
    Object {
      "colour": "red",
      "rotation": "UR",
      "rotation2": undefined,
      "type": "corner",
      "x": 7,
      "y": 4,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "DL",
      "rotation2": undefined,
      "type": "corner",
      "x": 9,
      "y": 4,
    },
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": undefined,
      "type": "corner",
      "x": 6,
      "y": 5,
    },
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    Object {
      "colour": "white",
      "rotation": "UR",
      "rotation2": undefined,
      "type": "corner",
      "x": 7,
      "y": 6,
    },
    null,
    null,
  ],
  Array [
    null,
    null,
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": undefined,
      "type": "corner",
      "x": 2,
      "y": 7,
    },
    Object {
      "colour": "white",
      "rotation": "U",
      "rotation2": undefined,
      "type": "block",
      "x": 3,
      "y": 7,
    },
    Object {
      "colour": "white",
      "rotation": undefined,
      "rotation2": undefined,
      "type": "king",
      "x": 4,
      "y": 7,
    },
    Object {
      "colour": "white",
      "rotation": "U",
      "rotation2": undefined,
      "type": "block",
      "x": 5,
      "y": 7,
    },
    null,
    null,
    null,
    Object {
      "colour": "white",
      "rotation": "U",
      "rotation2": undefined,
      "type": "laser",
      "x": 9,
      "y": 7,
    },
  ],
]
`;

exports[`getUserInput is called: Board state each turn 2`] = `
Array [
  Array [
    Object {
      "colour": "red",
      "rotation": "R",
      "rotation2": undefined,
      "type": "laser",
      "x": 0,
      "y": 0,
    },
    null,
    null,
    null,
    null,
    Object {
      "colour": "red",
      "rotation": undefined,
      "rotation2": undefined,
      "type": "king",
      "x": 5,
      "y": 0,
    },
    Object {
      "colour": "red",
      "rotation": "D",
      "rotation2": undefined,
      "type": "block",
      "x": 6,
      "y": 0,
    },
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": undefined,
      "type": "corner",
      "x": 7,
      "y": 0,
    },
    null,
    null,
  ],
  Array [
    null,
    null,
    Object {
      "colour": "red",
      "rotation": "DL",
      "rotation2": undefined,
      "type": "corner",
      "x": 2,
      "y": 1,
    },
    null,
    Object {
      "colour": "red",
      "rotation": "D",
      "rotation2": undefined,
      "type": "block",
      "x": 4,
      "y": 1,
    },
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    Object {
      "colour": "red",
      "rotation": "UR",
      "rotation2": undefined,
      "type": "corner",
      "x": 0,
      "y": 3,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "DL",
      "rotation2": undefined,
      "type": "corner",
      "x": 2,
      "y": 3,
    },
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": undefined,
      "type": "corner",
      "x": 3,
      "y": 3,
    },
    Object {
      "colour": "red",
      "rotation": "UR",
      "rotation2": "DL",
      "type": "mirror",
      "x": 4,
      "y": 3,
    },
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": "LU",
      "type": "mirror",
      "x": 5,
      "y": 3,
    },
    null,
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": undefined,
      "type": "corner",
      "x": 7,
      "y": 3,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": undefined,
      "type": "corner",
      "x": 9,
      "y": 3,
    },
  ],
  Array [
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": undefined,
      "type": "corner",
      "x": 0,
      "y": 4,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": undefined,
      "type": "corner",
      "x": 2,
      "y": 4,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": "RD",
      "type": "mirror",
      "x": 4,
      "y": 4,
    },
    Object {
      "colour": "white",
      "rotation": "DL",
      "rotation2": "UR",
      "type": "mirror",
      "x": 5,
      "y": 4,
    },
    null,
    Object {
      "colour": "red",
      "rotation": "UR",
      "rotation2": undefined,
      "type": "corner",
      "x": 7,
      "y": 4,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "DL",
      "rotation2": undefined,
      "type": "corner",
      "x": 9,
      "y": 4,
    },
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": undefined,
      "type": "corner",
      "x": 6,
      "y": 5,
    },
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    Object {
      "colour": "white",
      "rotation": "UR",
      "rotation2": undefined,
      "type": "corner",
      "x": 7,
      "y": 6,
    },
    null,
    null,
  ],
  Array [
    null,
    null,
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": undefined,
      "type": "corner",
      "x": 2,
      "y": 7,
    },
    Object {
      "colour": "white",
      "rotation": "U",
      "rotation2": undefined,
      "type": "block",
      "x": 3,
      "y": 7,
    },
    Object {
      "colour": "white",
      "rotation": undefined,
      "rotation2": undefined,
      "type": "king",
      "x": 4,
      "y": 7,
    },
    Object {
      "colour": "white",
      "rotation": "U",
      "rotation2": undefined,
      "type": "block",
      "x": 5,
      "y": 7,
    },
    null,
    null,
    null,
    Object {
      "colour": "white",
      "rotation": "U",
      "rotation2": undefined,
      "type": "laser",
      "x": 9,
      "y": 7,
    },
  ],
]
`;

exports[`getUserInput is called: Board state each turn 3`] = `
Array [
  Array [
    Object {
      "colour": "red",
      "rotation": "R",
      "rotation2": undefined,
      "type": "laser",
      "x": 0,
      "y": 0,
    },
    null,
    null,
    null,
    null,
    Object {
      "colour": "red",
      "rotation": undefined,
      "rotation2": undefined,
      "type": "king",
      "x": 5,
      "y": 0,
    },
    Object {
      "colour": "red",
      "rotation": "D",
      "rotation2": undefined,
      "type": "block",
      "x": 6,
      "y": 0,
    },
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": undefined,
      "type": "corner",
      "x": 7,
      "y": 0,
    },
    null,
    null,
  ],
  Array [
    null,
    null,
    Object {
      "colour": "red",
      "rotation": "DL",
      "rotation2": undefined,
      "type": "corner",
      "x": 2,
      "y": 1,
    },
    null,
    Object {
      "colour": "red",
      "rotation": "D",
      "rotation2": undefined,
      "type": "block",
      "x": 4,
      "y": 1,
    },
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    Object {
      "colour": "red",
      "rotation": "UR",
      "rotation2": undefined,
      "type": "corner",
      "x": 0,
      "y": 3,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "DL",
      "rotation2": undefined,
      "type": "corner",
      "x": 2,
      "y": 3,
    },
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": undefined,
      "type": "corner",
      "x": 3,
      "y": 3,
    },
    Object {
      "colour": "red",
      "rotation": "UR",
      "rotation2": "DL",
      "type": "mirror",
      "x": 4,
      "y": 3,
    },
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": "LU",
      "type": "mirror",
      "x": 5,
      "y": 3,
    },
    null,
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": undefined,
      "type": "corner",
      "x": 7,
      "y": 3,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": undefined,
      "type": "corner",
      "x": 9,
      "y": 3,
    },
  ],
  Array [
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": undefined,
      "type": "corner",
      "x": 0,
      "y": 4,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": undefined,
      "type": "corner",
      "x": 2,
      "y": 4,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": "RD",
      "type": "mirror",
      "x": 4,
      "y": 4,
    },
    Object {
      "colour": "white",
      "rotation": "DL",
      "rotation2": "UR",
      "type": "mirror",
      "x": 5,
      "y": 4,
    },
    null,
    Object {
      "colour": "red",
      "rotation": "UR",
      "rotation2": undefined,
      "type": "corner",
      "x": 7,
      "y": 4,
    },
    null,
    Object {
      "colour": "white",
      "rotation": "DL",
      "rotation2": undefined,
      "type": "corner",
      "x": 9,
      "y": 4,
    },
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    Object {
      "colour": "red",
      "rotation": "RD",
      "rotation2": undefined,
      "type": "corner",
      "x": 6,
      "y": 5,
    },
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    Object {
      "colour": "white",
      "rotation": "UR",
      "rotation2": undefined,
      "type": "corner",
      "x": 7,
      "y": 6,
    },
    null,
    null,
  ],
  Array [
    null,
    null,
    Object {
      "colour": "white",
      "rotation": "LU",
      "rotation2": undefined,
      "type": "corner",
      "x": 2,
      "y": 7,
    },
    Object {
      "colour": "white",
      "rotation": "U",
      "rotation2": undefined,
      "type": "block",
      "x": 3,
      "y": 7,
    },
    Object {
      "colour": "white",
      "rotation": undefined,
      "rotation2": undefined,
      "type": "king",
      "x": 4,
      "y": 7,
    },
    Object {
      "colour": "white",
      "rotation": "U",
      "rotation2": undefined,
      "type": "block",
      "x": 5,
      "y": 7,
    },
    null,
    null,
    null,
    Object {
      "colour": "white",
      "rotation": "U",
      "rotation2": undefined,
      "type": "laser",
      "x": 9,
      "y": 7,
    },
  ],
]
`;

exports[`getUserInput is called: Each turn 1`] = `"red"`;

exports[`getUserInput is called: Each turn 2`] = `"white"`;

exports[`getUserInput is called: Each turn 3`] = `"red"`;

exports[`getUserInput is called: Initial game 1`] = `
Object {
  "board": Array [
    Array [
      Object {
        "colour": "red",
        "rotation": "D",
        "rotation2": undefined,
        "type": "laser",
        "x": 0,
        "y": 0,
      },
      null,
      null,
      null,
      Object {
        "colour": "red",
        "rotation": "D",
        "rotation2": undefined,
        "type": "block",
        "x": 4,
        "y": 0,
      },
      Object {
        "colour": "red",
        "rotation": undefined,
        "rotation2": undefined,
        "type": "king",
        "x": 5,
        "y": 0,
      },
      Object {
        "colour": "red",
        "rotation": "D",
        "rotation2": undefined,
        "type": "block",
        "x": 6,
        "y": 0,
      },
      Object {
        "colour": "red",
        "rotation": "RD",
        "rotation2": undefined,
        "type": "corner",
        "x": 7,
        "y": 0,
      },
      null,
      null,
    ],
    Array [
      null,
      null,
      Object {
        "colour": "red",
        "rotation": "DL",
        "rotation2": undefined,
        "type": "corner",
        "x": 2,
        "y": 1,
      },
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
      Object {
        "colour": "white",
        "rotation": "LU",
        "rotation2": undefined,
        "type": "corner",
        "x": 3,
        "y": 2,
      },
      null,
      null,
      null,
      null,
      null,
      null,
    ],
    Array [
      Object {
        "colour": "red",
        "rotation": "UR",
        "rotation2": undefined,
        "type": "corner",
        "x": 0,
        "y": 3,
      },
      null,
      Object {
        "colour": "white",
        "rotation": "DL",
        "rotation2": undefined,
        "type": "corner",
        "x": 2,
        "y": 3,
      },
      null,
      Object {
        "colour": "red",
        "rotation": "UR",
        "rotation2": "DL",
        "type": "mirror",
        "x": 4,
        "y": 3,
      },
      Object {
        "colour": "red",
        "rotation": "RD",
        "rotation2": "LU",
        "type": "mirror",
        "x": 5,
        "y": 3,
      },
      null,
      Object {
        "colour": "red",
        "rotation": "RD",
        "rotation2": undefined,
        "type": "corner",
        "x": 7,
        "y": 3,
      },
      null,
      Object {
        "colour": "white",
        "rotation": "LU",
        "rotation2": undefined,
        "type": "corner",
        "x": 9,
        "y": 3,
      },
    ],
    Array [
      Object {
        "colour": "red",
        "rotation": "RD",
        "rotation2": undefined,
        "type": "corner",
        "x": 0,
        "y": 4,
      },
      null,
      Object {
        "colour": "white",
        "rotation": "LU",
        "rotation2": undefined,
        "type": "corner",
        "x": 2,
        "y": 4,
      },
      null,
      Object {
        "colour": "white",
        "rotation": "LU",
        "rotation2": "RD",
        "type": "mirror",
        "x": 4,
        "y": 4,
      },
      Object {
        "colour": "white",
        "rotation": "DL",
        "rotation2": "UR",
        "type": "mirror",
        "x": 5,
        "y": 4,
      },
      null,
      Object {
        "colour": "red",
        "rotation": "UR",
        "rotation2": undefined,
        "type": "corner",
        "x": 7,
        "y": 4,
      },
      null,
      Object {
        "colour": "white",
        "rotation": "DL",
        "rotation2": undefined,
        "type": "corner",
        "x": 9,
        "y": 4,
      },
    ],
    Array [
      null,
      null,
      null,
      null,
      null,
      null,
      Object {
        "colour": "red",
        "rotation": "RD",
        "rotation2": undefined,
        "type": "corner",
        "x": 6,
        "y": 5,
      },
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      Object {
        "colour": "white",
        "rotation": "UR",
        "rotation2": undefined,
        "type": "corner",
        "x": 7,
        "y": 6,
      },
      null,
      null,
    ],
    Array [
      null,
      null,
      Object {
        "colour": "white",
        "rotation": "LU",
        "rotation2": undefined,
        "type": "corner",
        "x": 2,
        "y": 7,
      },
      Object {
        "colour": "white",
        "rotation": "U",
        "rotation2": undefined,
        "type": "block",
        "x": 3,
        "y": 7,
      },
      Object {
        "colour": "white",
        "rotation": undefined,
        "rotation2": undefined,
        "type": "king",
        "x": 4,
        "y": 7,
      },
      Object {
        "colour": "white",
        "rotation": "U",
        "rotation2": undefined,
        "type": "block",
        "x": 5,
        "y": 7,
      },
      null,
      null,
      null,
      Object {
        "colour": "white",
        "rotation": "U",
        "rotation2": undefined,
        "type": "laser",
        "x": 9,
        "y": 7,
      },
    ],
  ],
  "message": "",
  "turn": "red",
}
`;
